Add storage RB config.

Pchelolo opened the pull request:
Bug: [T220855](https://phabricator.wikimedia.org/T220855)

d00rman commented:
Very early WIP, but yeah, this is the way to go. One thing we'll have to definitely change is using the `/sys` hierarchy. Since we will still need it in the front-end service, here we will probably just expose all of this via `/v1`. That is safe to do since the back-end service will not be exposed to the public. Oh and we need to figure out the names for these two services.

Pchelolo commented:
Sure sure, it's not ready for anything in any way, I've just started playing with this and decided to publish my playground early to have a venue for occasional bikesheds.

I actually like the idea of converting it to `/v1` - that would mean we wouldn't need to unrestrict the `/sys` that adds a layer of protection for the public RB instance.

Pchelolo commented:
heh.. it's tricky... the paths to the buckes are used internally under `/sys`...

d00rman commented:
> heh.. it's tricky... the paths to the buckes are used internally under `/sys`...

We could mount a simple proxy module in the config to redirect the requests to the back-end, something like:

```yaml
/{api:sys}:
  x-modules:
    - spec:
        paths:
          /page_revisions:
            x-modules:
              - path: sys/page_revisions.js
          /post_data: 
            x-modules:
              - path: sys/post_data.js
          /action:
            x-modules:
              - path: sys/action.js
                options: "{{options.action}}"
          /page_save:
            x-modules:
              - path: sys/page_save.js
          /parsoid:
            x-modules:
              - path: sys/parsoid.js
                options: &parsoid_options
                  parsoidHost: '{{options.parsoid.host}}'
                  response_cache_control: '{{options.purged_cache_control}}'
                  grace_ttl: '{{default(options.parsoid.grace_ttl, 86400)}}'
          /{rest}:
            x-modules:
              - path: sys/proxy.js
                options:
                  backend_host: 127.0.0.1:7234
```

where `sys/proxy.js` would then send the request to the back-end service.

Pchelolo commented:
Basically, in the new version, I've switched summary to work in multi-instance configuration and switched one test suite to pass in the multi-configuration settings. (still need to push a bit of an update to hyperswitch), but I think this is basically a demonstration of how this whole thing will work.

The servers we start up for different CIs will be controlled by env variables.

Pchelolo commented:
So please have a look before I start working on it more

d00rman commented on config.example.storage.wikimedia.yaml:
```diff
@@ -0,0 +1,113 @@
+                            dependent_content_types:
+                              - name: data-parsoid
+                                value_type: json
+                    /key_value:
```
I think we'll need here only `key_value` and `page_revisions` once the simplification is completed.

Pchelolo replied:
Yes. But it's not complete yet, so I needed this to make it work. It will go away.

d00rman commented on config.example.storage.wikimedia.yaml:
```diff
@@ -0,0 +1,113 @@
+  name: restbase
+
+services:
+  - name: restbase
```
We probably want to chnge the name(s)

Pchelolo replied:
to something we didn't decide on yet.

d00rman replied:
Heh, we've settled on restbase, so no need to change :)

d00rman commented on config.example.storage.wikimedia.yaml:
```diff
@@ -0,0 +1,113 @@
+#  port: 8125
+#  batch: true
+
+ratelimiter:
```
Would we need rate limiting for storage given that it's not going to be exposed?

Pchelolo replied:
indeed we don't

d00rman replied:
Still not needed :P

d00rman commented on config.example.storage.wikimedia.yaml:
```diff
@@ -0,0 +1,112 @@
+  paths:
+    /{domain}: *default_project
+    # A robots.txt to make sure that the content isn't indexed.
+    /robots.txt:
```
`robots.txt` will be matched by `/{domain}` above, so it needs to be listed before.

Pchelolo replied:
Actually, it needs to be removed entirely. This is not going to be exposed outside.

d00rman replied:
Let's not do that. Let's minimise the changes we make outside the split.

Pchelolo replied:
mmm? This is a completely new config, so I don't get how not removing this would minimize any of the changes. It's a copy-paste artifact.

d00rman replied:
No, it's not. It's the front-end component of the current RB. And the current status quo is that RESTBase-the-router exposes it. So let's keep it for now.

Pchelolo replied:
This is not the frontend component :) This is all complicated, but this is the example of how the backend component will look like when we finish up with the transition :)

d00rman replied:
/me facepalms. Dang you are right. And it says nicely in the file name lol. Ok, yes, let's remove it.

d00rman commented on sys/backend_proxy.js:
```diff
@@ -0,0 +1,45 @@
+                    query: req.query
+                })
+                .then((res) => {
+                    // TODO: Leave a comment why.
```
Yes, please do :)

d00rman commented on sys/backend_proxy.js:
```diff
@@ -0,0 +1,45 @@
+        },
+        operations: {
+            proxy: (hyper, req) => {
+                // Add the proxied module name to the path.
```
Could you add a slightly-longer desc and an example for easier comprehension, please?

d00rman commented on test/utils/run_tests.sh:
```diff
@@ -22,30 +27,47 @@ runTest ( ) {
-        "${mocha}"
-    elif [ "$2" = "coverage" ]; then
-        "${nyc}" --reporter=lcov node_modules/.bin/_mocha
+    if [ "x$3" = "x" ] || [ "$3" = "single_process" ]; then
```
For future compatibility with WMF CI runs, it would be good for both `$2` and `$3` to be able to be set via `TEST_TARGET` and `TEST_MODE` (and for env vars to be preferred over the positional args).

Pchelolo replied:
I was thinking about this, and I think we need to just remove the test runner script. Fallback to SQLite is default now via the default value of the RB_TEST_BACKEND env variable in the config. Locally running all configurations is unlikely to happen a lot, in Travis we can achieve running all the configs via env variables matrix. I think I'll just document the env variables and remove the run_tests.sh. What do you think?

Pchelolo replied:
Doing that.

d00rman replied:
I actually liked your earlier idea about removing this altogether and control stuff only with ENV VARS.

d00rman commented on config.example.wikimedia.yaml:
```diff
@@ -6,6 +6,8 @@ default_project: &default_project
   x-modules:
     - path: projects/wmf_default.yaml
       options: &default_options
+        proxy:
```
It would be good to also have an example config just for RESTRouter, since this file now contains a lot of unused config options.

Pchelolo replied:
Only the `table` configuration is unused now. Do we want to create a copy-paste of a 200 lines long file just to make the example?

d00rman replied:
But most of the config is put in there, so I think it'll be clearer that way

Pchelolo replied:
I'm not quite sure I follow regarding 'most of the config'

d00rman replied:
Euh, I meant most of the config vars that, say, come from puppet are declared here.

d00rman replied:
Nm...

Pchelolo commented on package.json:
```diff
@@ -37,6 +37,7 @@
   "dependencies": {
     "bluebird": "^3.5.5",
     "cassandra-uuid": "^0.1.0",
+    "colors": "^1.3.3",
```
Ups...

Pchelolo commented on sys/parsoid.js:
```diff
@@ -77,7 +77,7 @@ function _dependenciesUpdate(hyper, req, newContent = true) {
     return mwUtil.getSiteInfo(hyper, req)
     .then((siteInfo) => {
-        const baseUri = siteInfo.baseUri.replace(/^https?:/, '');
+        const baseUri = siteInfo.baseUri.toString().replace(/^https?:/, '');
```
Not needed anymore.

d00rman commented:
Very well done! Some more things I think should be done further, though:

- the projects shouldn't contain `/sys/table` and friends. We will use the projects in RESTRouter, so they shouldn't be loaded at all. As we discussed, in the interim we should have two versions of the projects: one for the whole vertical and one just for the front-end
- for completeness, it would be good imho to also have an example config for the front-end

d00rman commented:
Superseded by #1148
